presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/StreamPreferredProperties.java

```diff
@@ -0,0 +1,235 @@
+    }
+
+    public static StreamPreferredProperties singleStream()
+    {
+        return new StreamPreferredProperties(Optional.of(SINGLE), Optional.empty(), false);
+    }
+
+    public static StreamPreferredProperties fixedParallelism()
+    {
+        return new StreamPreferredProperties(Optional.of(FIXED), Optional.empty(), false);
+    }
+
+    public static StreamPreferredProperties defaultParallelism(Session session)
+    {
+        if (getTaskConcurrency(session) > 1 && !preferStreamingOperators(session)) {
+            return new StreamPreferredProperties(Optional.of(MULTIPLE), Optional.empty(), false);
+        }
+        return any();
+    }
+
+    public StreamPreferredProperties withParallelism()
+    {
+        // do not override an existing parallel preference
+        if (isParallelPreferred()) {
+            return this;
+        }
+        return new StreamPreferredProperties(Optional.of(MULTIPLE), Optional.empty(), orderSensitive);
+    }
+
+    public StreamPreferredProperties withoutPreference()
+    {
+        return new StreamPreferredProperties(Optional.empty(), Optional.empty(), orderSensitive);
+    }
+
+    public StreamPreferredProperties withPartitioning(Collection<Symbol> partitionSymbols)
+    {
+        if (partitionSymbols.isEmpty()) {
+            return singleStream();
+        }
+
+        Iterable<Symbol> desiredPartitioning = partitionSymbols;
+        if (partitioningColumns.isPresent()) {
+            // If there are common columns between our requirements and the desired partitionSymbols, both can be satisfied in one shot
+            Set<Symbol> common = Sets.intersection(ImmutableSet.copyOf(desiredPartitioning), ImmutableSet.copyOf(partitioningColumns.get()));
+
+            // If we find common partitioning columns, use them, else use child's partitioning columns
+            if (!common.isEmpty()) {
+                desiredPartitioning = common;
+            }
+        }
+
+        return new StreamPreferredProperties(distribution, Optional.of(desiredPartitioning), false);
+    }
+
+    public StreamPreferredProperties withDefaultParallelism(Session session)
+    {
+        if (getTaskConcurrency(session) > 1 && !preferStreamingOperators(session)) {
+            return withParallelism();
+        }
+        return this;
+    }
+
+    public boolean isSatisfiedBy(StreamProperties actualProperties)
+    {
+        // is there a specific preference
+        if (!distribution.isPresent() && !partitioningColumns.isPresent()) {
+            return true;
+        }
+
+        if (isOrderSensitive() && actualProperties.isOrdered()) {
+            // ordered is required to be a single stream, so in this ordered case SINGLE is
+            // considered satisfactory for MULTIPLE and FIXED
+            return true;
+        }
+
+        if (distribution.isPresent()) {
+            StreamDistribution actualDistribution = actualProperties.getDistribution();
+            switch (distribution.get()) {
```

Comment:
I think this would be easier to read as an `if`:

```java
if (distribution.get() == SINGLE && actualDistribution != SINGLE) {
    return false;
}

if (distribution.get() == FIXED && actualDistribution != FIXED) {
    return false;
}

if (distribution.get() == MULTIPLE && actualDistribution != FIXED && actualDistribution != MULTIPLE) {
    return false;
}
```